Implement ExternalWork for ESProducers

wddgit opened the pull request:
#### PR description:

Implement ExternalWork for ESProducers

This new feature is analogous to the ExternalWork ability already implemented for EDProducers and EDFilters. It behaves in almost the same way. There is an acquire function that runs that can get data from the EventSetup, then the ExternalWork runs asynchronously, and finally a produce method runs. The ExternalWork has a WaitingTaskWithArenaHolder to let the produce method know when it can start. It is important that the acquire method completes and is no longer on the stack while the ExternalWork is running.

The interface to use this feature is provided by a new base class named ESProducerExternalWork that derives from the ESProducer class. It has new functions named setWhatAcquiredProduced (6 functions with this name) and setWhatAcquiredProducedWithLambda (2 functions with this name). To utilize the new feature, use one of the new functions instead of one of the old functions named setWhatProduced. The new acquire function must be provided to the ESProducerExternalWork in a way completely analogous to the way one supplies the produce function, with all the same alternatives. If a class has multiple produce methods, some can use the new interface and some can use the existing interface which is unchanged.

For example, one could define the following function as a member function of the class deriving from ESProducerExternalWork.

```
void acquire(MyRecordType const&, edm::WaitingTaskWithArenaHolder);
```

Or one could provide a lambda as an argument to the function call. Or one could define a similar function with a different name and provide a pointer to that function as an argument to the function. 

The behavior of existing code that uses the older setWhatProduced functions should not change. That interface is not modified, although there are some changes in the underlying implementation details.

@fwyzard This adds two new signals to the ActivityRegistry that you might or might not want to include in the fast timer service.

(This is my second attempt at this. See #40798 for the first attempt and review comments. Since that attempt I refactored the Callback class to 3 classes, CallbackBase, CallBack, and CallbackExternalWork and refactored ESProducer into 2 separate classes ESProducer and ESProducerExternalWork. The code moved around so much it seemed easier to start with a fresh PR. I extended the new test so that the test ESProducer has two "produce" functions, one with an acquire function and one without. In terms of behavior and interface, nothing important changed other than ESProducers using the new feature should inherit from ESProducerExternalWork instead of ESProducer.)

We discussed an additional feature that was related to the acquire function returning a value other than void. We decided to postpone that development for a separate PR to be submitted later.

#### PR validation:

At this point nothing uses this new feature other than new unit tests in the Framework added by this PR. These new unit tests pass. Existing code should behave identically so this relies on existing tests for ESProducers not using using the new ExternalWork feature.

wddgit committed:
Implement ExternalWork for ESProducers

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41001/34517

- This PR adds an extra 92KB to repository

cmsbuild commented:
A new Pull Request was created by  @wddgit (W. David Dagenhart) for master.

It involves the following packages:

- FWCore/Framework (**core**)
- FWCore/Integration (**core**)
- FWCore/ServiceRegistry (**core**)
- FWCore/Services (**core**)
- FWCore/Utilities (**core**)


@cmsbuild, @smuzaffar, @Dr15Jones, @makortel can you please review it and eventually sign? Thanks.
@makortel, @missirol, @fwyzard, @felicepantaleo this is something you requested to watch as well.
@perrotta, @dpiparo, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

wddgit commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-46a227/31150/summary.html
**COMMIT**: 8f8746cf3b2b35b407470e262b782de0d0ad4ab8
**CMSSW**: CMSSW_13_1_X_2023-03-08-1300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/41001/31150/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially removed 1 lines from the logs
 - Reco comparison results: 6 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3530381
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3530353
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 213 log files, 164 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-03-08-1300+46a227/55909/triggerResults): no differences found

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41001/34589

- This PR adds an extra 20KB to repository

wddgit commented:
Rebased to remove conflict

cmsbuild commented:
Pull request #41001 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @makortel can you please check and sign again.

wddgit commented on FWCore/Framework/interface/CallbackBase.h:
```diff
@@ -0,0 +1,270 @@
+      const CallbackBase& operator=(const CallbackBase&) = delete;
+      virtual ~CallbackBase() = default;
+
+      virtual CallbackBase* clone() = 0;
```
I agree. Having it virtual is unnecessary. I changed this (I'll push all the commits together after I've worked through all the comments).

wddgit commented on FWCore/Framework/interface/CallbackBase.h:
```diff
@@ -0,0 +1,270 @@
+        if (doPrefetch) {
+          auto group = iTask.group();
+          ServiceWeakToken weakToken(token);
+          WaitingTaskHolder runModuleTaskHolder = makeRunModuleTaskHolder(group, weakToken, iRecord, iEventSetupImpl);
```
@makortel I tried the Curiously Recurring Template Pattern. It's the third commit. A little extra complexity. Possibly premature optimization. I'm happy with it though. It was only a little extra complexity and was interesting to implement... It is the third commit if you want to take a look. I just pushed it.

makortel replied:
I agree CRTP is possibly a premature optimization. My main motivation was that the related code was already always dealing with the concrete `Callback` type without any virtual functions, and I wanted to preserve that (as long as it is feasible).

wddgit replied:
I pushed the third commit right away because I wanted some feedback on that one. Do you want me to back the CRTP version out and go back to a using a virtual function? Both ways work and it is a trade off (a virtual function is slightly simpler for a human code reader, CRTP probably runs slightly faster). I'm OK either way, just let me know which you prefer.

(As for the rest, I was still working on revisions for the comments. I'll push new commits soon. I got sidetracked on the office move yesterday and then my home WiFi breaking this morning. I spent half the morning talking to COMCAST... It is back up. I hope it stays that way.)

Dr15Jones replied:
I have a 3rd suggestion which avoids the virtual function (which I'd prefer we do not use) and CRTP. You could create a protected function in the base class that takes a templated pointer to the actual derived type,
```
template<typename DERIVED>
WaitingTaskHolder prefetchAsyncImpl(DERIVED* derived, ....)
```

Then require inheriting classes define the regular `prefetchAsync` which just calls `prefetchAsyncImpl` with the `this` pointer added. We do something similar with WorkerT and Worker.

Dr15Jones replied:
Hmm, if the only thing it needs is the value returned from `makeRunModuleTaskHolderImpl` then maybe it would be even simpler just to have

```
template<typename RunModuleFnctr>
WaitingTaskHolder prefetchAsyncImpl(RunModuleFnctr&&, ...)
```

and then prefetchAsync is just
```
WaitingTaskHolder prefetchAsync(...) {
   return prefetchAsyncImpl( [this](auto&& group, auto&& token, auto&& record, auto&& es) { return makeRunModuleTaskHolder(group, token, record, es);}, ...);
```

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41001/34609

- This PR adds an extra 32KB to repository

cmsbuild commented:
Pull request #41001 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @makortel can you please check and sign again.

wddgit commented on FWCore/Framework/interface/CallbackExternalWork.h:
```diff
@@ -0,0 +1,193 @@
+    public:
+      using Base = CallbackBase<T, TProduceFunc, TReturn, TRecord, TDecorator>;
+
+      CallbackExternalWork(T* iProd, TProduceFunc iProduceFunc, unsigned int iID, const TDecorator& iDec = TDecorator())
```
Thanks for noticing this. You are right. This is harmless, but serves no purpose. If there is not an acquire function, then the Callback class should always be used. There was a little bit of related logic in the makeRunModuleTaskHolderImpl function that could also be removed, so I removed it. This was leftover from before I refactored the class.

makortel commented:
@cmsbuild, please test

makortel commented on FWCore/Framework/interface/CallbackExternalWork.h:
```diff
@@ -0,0 +1,199 @@
+                                                    ServiceWeakToken const& token,
+                                                    EventSetupRecordImpl const* eventSetupRecordImpl,
+                                                    EventSetupImpl const* eventSetupImpl) {
+        bool emitPostPrefetchingSignal = true;
+        if (acquireFunction_) {
+          emitPostPrefetchingSignal = false;
+        }
```
Now that `acquireFunction_` should always be set, could this become
```suggestion
        constexpr bool emitPostPrefetchingSignal = false;
```

wddgit replied:
I agree. Done.

makortel commented on FWCore/Framework/interface/CallbackExternalWork.h:
```diff
@@ -0,0 +1,199 @@
+        if (acquireFunction_) {
+          emitPostPrefetchingSignal = false;
+        }
+        WaitingTask* runModuleTask =
+            Base::makeProduceTask(group, token, eventSetupRecordImpl, eventSetupImpl, emitPostPrefetchingSignal);
+
+        if (acquireFunction_) {
+          WaitingTaskWithArenaHolder waitingTaskWithArenaHolder(*group, makeExceptionHandlerTask(runModuleTask, group));
+          runModuleTask = makeAcquireTask(
+              std::move(waitingTaskWithArenaHolder), group, token, eventSetupRecordImpl, eventSetupImpl);
+        }
```
I think these tasks could also be set up under the assumption that `acquireFunction_` is always set.

wddgit replied:
Done.

makortel commented on FWCore/Framework/interface/CallbackExternalWork.h:
```diff
@@ -0,0 +1,199 @@
+        });
+      }
+
+      std::shared_ptr<TAcquireFunc> acquireFunction_;
```
Should this member data be `private:'? (I didn't notice one) Maybe also some of the member functions?

wddgit replied:
Yes. Done. I made everything private that could be made private.

makortel commented on FWCore/Integration/plugins/AcquireIntESProducer.cc:
```diff
@@ -0,0 +1,152 @@
+  }
+
+  std::unique_ptr<ESTestDataB> AcquireIntESProducer::produceESTestDataB(ESTestRecordB const&) {
+    auto testDataB = std::make_unique<ESTestDataB>(11);
+    return testDataB;
```
How about just
```suggestion
    return std::make_unique<ESTestDataB>(11);
```
?

wddgit replied:
Done. Better. Thanks.

makortel commented on FWCore/Integration/plugins/AcquireIntESProducer.cc:
```diff
@@ -0,0 +1,152 @@
+      : numberOfIOVsToAccumulate_(pset.getUntrackedParameter<unsigned int>("numberOfIOVsToAccumulate")),
+        secondsToWaitForWork_(pset.getUntrackedParameter<unsigned int>("secondsToWaitForWork")) {
+    auto collector = setWhatAcquiredProduced(this, "fromAcquireIntESProducer");
+    setWhatProduced(this, &edmtest::AcquireIntESProducer::produceESTestDataB);
+    token_ = collector.consumes<IOVTestInfo>(edm::ESInputTag{"", ""});
```
I would reverse these two lines
```suggestion
    token_ = collector.consumes<IOVTestInfo>(edm::ESInputTag{"", ""});
    setWhatProduced(this, &edmtest::AcquireIntESProducer::produceESTestDataB);
```
because the `token_` is related to the functions registered by `setWhatAcquiredProduced()`.

wddgit replied:
Done. Better.

Dr15Jones commented on FWCore/Framework/interface/CallbackBase.h:
```diff
@@ -0,0 +1,270 @@
+
+      CallbackBase(const CallbackBase&) = delete;
+      const CallbackBase& operator=(const CallbackBase&) = delete;
+      virtual ~CallbackBase() = default;
```
The destructor is still virtual.

wddgit replied:
I just pushed it. It was still in my working area. Done now.

Dr15Jones commented on FWCore/Framework/interface/CallbackExternalWork.h:
```diff
@@ -0,0 +1,199 @@
+                           TProduceFunc iProduceFunc,
+                           unsigned int iID,
+                           const TDecorator& iDec = TDecorator())
+          : Base(iProd, std::make_shared<TProduceFunc>(std::move(iProduceFunc)), iID, iDec),
```
How about calling the constructor just below this one instead (i.e. it is then just a forwarding constructor).

wddgit replied:
Done.

Dr15Jones commented on FWCore/Framework/interface/CallbackExternalWork.h:
```diff
@@ -0,0 +1,199 @@
+      }
+
+      WaitingTask* makeExceptionHandlerTask(WaitingTask* produceTask, oneapi::tbb::task_group* group) {
+        return make_waiting_task([this, produceTask, group](std::exception_ptr const* iException) {
```
I'm not a fan of the `produceTask` being held by a 'bare' pointer. I'd rather have `produceTask = WaitingTaskHolder(*group, produceTask)` as that will increment the reference count and guarantee the task will be queued even if the task is deleted without ever being run directly.

Dr15Jones replied:
Or maybe just change `WaitingTask* produceTask` to `WaitingTaskHolder produceTask`.

wddgit replied:
I modified the make functions to return WaitingTaskHolders instead of WaitingTask*. The other way made sense before I refactored. But now this is much better.

Dr15Jones commented on FWCore/Framework/interface/ESProducer.h:
```diff
@@ -238,11 +234,19 @@ namespace edm {
+    // other classes derived from them.
+    unsigned int consumesInfoSize() const { return consumesInfos_.size(); }
+
+    void consumesInfoPushBackNew() { consumesInfos_.push_back(std::make_unique<ESConsumesInfo>()); }
```
What if this returned the entry added? Then you wouldn't need `consumesInfoPointerToLastInfo()`.

wddgit replied:
Done.

Dr15Jones commented on FWCore/Integration/test/testESProducerUsingAcquire_cfg.py:
```diff
@@ -0,0 +1,69 @@
+    numberEventsInRun = cms.untracked.uint32(100)
+)
+
+process.maxEvents = cms.untracked.PSet(
+    input = cms.untracked.int32(8)
+)
```
You mean `process.maxEvents.input = 8`

wddgit replied:
Done. I always forget that new syntax and copy from an old config. Maybe we should update all the old ones...

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-46a227/31297/summary.html
**COMMIT**: a089a92d495be07582cad50704844525c81c4a38
**CMSSW**: CMSSW_13_1_X_2023-03-14-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/41001/31297/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found errors in the following unit tests:

<pre>---> test CalibCalorimetryEcalLaserSortingRunStreamer had ERRORS
---> test DQMServicesStreamerIORunStreamer had ERRORS
</pre>

## Comparison Summary



**Summary**:
 - You potentially removed 17 lines from the logs
 - Reco comparison results: 12 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3550756
 - DQMHistoTests: Total failures: 7
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3550727
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 213 log files, 164 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-03-14-2300+46a227/56033/triggerResults): no differences found

makortel commented:
The unit test failures are
```
===== Test "DQMServicesStreamerIORunStreamer" ====
LOCAL_TEST_DIR = /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_13_1_X_2023-03-14-2300/src/DQMServices/StreamerIO/test
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_13_1_X_2023-03-14-2300/src/DQMServices/StreamerIO/test/RunStreamer.sh: line 15: out: Is a directory
Failure cmsRun streamOut_cfg.py: status 1
```
```
===== Test "CalibCalorimetryEcalLaserSortingRunStreamer" ====
LOCAL_TEST_DIR = /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_13_1_X_2023-03-14-2300/src/CalibCalorimetry/EcalLaserSorting/test
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_13_1_X_2023-03-14-2300/src/CalibCalorimetry/EcalLaserSorting/test/RunStreamer.sh: line 13: out: Is a directory
Failure cmsRun streamOut_cfg.py: status 1
```
I've noticed these in the IBs as well (although not in all flavors). @smuzaffar Could these be related to the unit testing improvements you have been doing recently?

smuzaffar commented:
@makortel , https://github.com/cms-sw/cmssw/pull/41055 should fix these unit tests. But as I just have merged https://github.com/cms-sw/cmsdist/pull/8385 ( to enable all unit tests use separate dir) so #41055 is not needed any more. Next IBs should fix the unit test issue

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41001/34692

- This PR adds an extra 140KB to repository

wddgit commented on FWCore/Framework/interface/CallbackExternalWork.h:
```diff
@@ -0,0 +1,193 @@
+            try {
+              convertException::wrap([excptr]() { std::rethrow_exception(excptr); });
+            } catch (cms::Exception& exception) {
+              exception.addContext("Running acquire and external work");
```
It is the first thing that can get the exception from the ExternalWork and maybe from acquire if it got far enough. It can convert to cms::Exception, add the usual context, and acquire specific context.

cmsbuild commented:
Pull request #41001 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @makortel can you please check and sign again.

wddgit commented on FWCore/Framework/interface/ESProducer.h:
```diff
@@ -238,11 +234,19 @@ namespace edm {
+
+    void consumesInfoPushBackNew() { consumesInfos_.push_back(std::make_unique<ESConsumesInfo>()); }
+
+    ESConsumesInfo* consumesInfoPointerToLastInfo() { return consumesInfos_.back().get(); }
```
Done. I used Chris's suggestion below.

wddgit commented on FWCore/Framework/interface/ESProducerExternalWork.h:
```diff
@@ -0,0 +1,146 @@
+      return setWhatAcquiredProduced(iThis, es::Label(iLabel));
+    }
+    template <typename T>
+    auto setWhatAcquiredProduced(T* iThis, const std::string& iLabel) {
```
I think this should be in a separate PR. It probably affects es::Label and ESProducer.h... Should I make an issue?

makortel replied:
I agree a general change towards `std::string_view` here would be better done in a separate PR. Let's go now with an issue.

wddgit replied:
See issue #41086.

wddgit commented on FWCore/Utilities/src/Exception.cc:
```diff
@@ -174,9 +178,12 @@ namespace cms {
 
-  void Exception::setAlreadyPrinted(bool value) { alreadyPrinted_ = value; }
+  bool Exception::setAlreadyPrinted() {
+    bool expected = false;
```
Done.

wddgit commented:
Thanks for all the comments. The PR is better because of them.

I would like to squash before this gets merged, but at the moment the fixes are in separate commits. Maybe that helps to review... Let me know when I can squash things.

I think I responded to all the comments. Let me know if there are more or if I missed something.

wddgit commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-46a227/31351/summary.html
**COMMIT**: 23c5afd1df201266f05dc4eb4169bd9a3d3e21f5
**CMSSW**: CMSSW_13_1_X_2023-03-16-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/41001/31351/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially added 8 lines to the logs
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3550819
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3550791
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 213 log files, 164 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-03-16-1100+46a227/56066/triggerResults): no differences found

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41001/34709

- This PR adds an extra 24KB to repository

cmsbuild commented:
Pull request #41001 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @makortel can you please check and sign again.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-46a227/31384/summary.html
**COMMIT**: e030f0dec68d90cb012c0da88ba15f69eb78a785
**CMSSW**: CMSSW_13_1_X_2023-03-17-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/41001/31384/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



There are some workflows for which there are errors in the baseline:
 24234.61 step 2
 The results for the comparisons for these workflows could be incomplete 
This means most likely that the IB is having errors in the relvals.The error does NOT come from this pull request

**Summary**:
 - You potentially removed 8 lines from the logs
 - Reco comparison results: 11 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3550882
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3550854
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 213 log files, 164 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-03-17-1100+46a227/56088/triggerResults): no differences found

wddgit commented on FWCore/Framework/interface/Callback.h:
```diff
@@ -9,231 +9,80 @@
-            rec->prefetchAsync(task, proxies[i], iImpl, token, edm::ESParentContext{&callingContext_});
-          }
-        }
+      Callback(T* iProd,
```
One thing I just noticed, I made this constructor private and the other constructor now delegates to it because of comments from Matti (a general comment about making things private that could be private) and Chris (about delegating instead directly initializing the base class), but perhaps a better modification would be to delete this constructor. Nothing uses it. I only included it because it existed before. Is there some expectation that we will need this in the future or that code outside of the repository needs this? It's a very minor thing either way.

Also let me know when the review is complete and I'll squash all these commits.

makortel replied:
> but perhaps a better modification would be to delete this constructor. Nothing uses it.

Isn't this constructor being used by `clone()`? (or did I miss something?)

> Also let me know when the review is complete and I'll squash all these commits.

I want to take another look after your latest changes, I hope to get to it today.

wddgit replied:
You're right. Sorry for the noise. It is used by clone.

makortel reviewed (commented):
Chris added the ESModule signal use to `Timing` service recently in https://github.com/cms-sw/cmssw/pull/41052. I think the `Timing` service should be extended to account the time spent in `{pre,post}ESModuleAquire` as well, but maybe that is better to be left for a subsequent PR?

wddgit commented on FWCore/Framework/interface/CallbackBase.h:
```diff
@@ -0,0 +1,270 @@
+            decorator_(iDec) {}
+
+      CallbackBase(const CallbackBase&) = delete;
+      const CallbackBase& operator=(const CallbackBase&) = delete;
```
Hmmmm.

You are certainly correct about the return type. I changed it and will push it into GitHub soon (I'll push all 4 commits at the same time when they are done). I had copied/pasted this from Callback.h and just changed the name without even thinking... Note that this error occurs often in FWCore/Framework. I didn't test but it might even be in the FWCore/Skeletons code generation. It is rare for this to make any practical difference, but it is incorrect.

About the move operators. I am leaning towards leaving them out. Deleting the copy constructor/assignment also deletes the move operators unless you explicitly define them so it seems like unnecessary boilerplate to me. In FWCore/Framework the vast majority of the time that is the pre-existing pattern. The rule of five in the CMS Style Guide says  "definition of any of the following requires definition of all five", but deletion is not definition... I don't feel strongly about this.

makortel replied:
I interpret the rule of five such that deletion is one kind of definition. I find the rules for compiler-generated copy/move operators complicated, and prefer to define all of them explicitly (if any of them need to be defined) so that the intended behavior is clear. Although in practice the `std::atomic<bool>` member effectively disables all copy and move operations anyway.

wddgit replied:
Done. I followed your interpretation of the rule of 5. I also cleaned up the other classes I significantly modified in this PR related to this and the assignment return value. (There are lots of other files in the Framework that have the same issues.)

makortel replied:
Thanks!

> (There are lots of other files in the Framework that have the same issues.)

I'd bet many of them predate C++11 and the move construction/assignment.

wddgit commented on FWCore/Framework/interface/Callback.h:
```diff
@@ -9,231 +9,80 @@
-        setData<typename type::head_type, typename type::tail_type>(iReturn);
+        return Base::prefetchAsyncImpl(
+            [this](auto&& group, auto&& token, auto&& record, auto&& es) {
+              return makeRunModuleTaskHolder(group, token, record, es);
```
Done. More concise. Thanks.

wddgit commented:
I'll rebase to a release that includes the Timing PR from Chris and include that also. I'm presuming you only want to add in the time from acquire and not the time of the ExternalWork. It looks fairly easy.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41001/34763

- This PR adds an extra 20KB to repository

cmsbuild commented:
Pull request #41001 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @makortel can you please check and sign again.

makortel commented:
> I'll rebase to a release that includes the Timing PR from Chris and include that also.

Ok.

> I'm presuming you only want to add in the time from acquire and not the time of the ExternalWork

Correct, addition of only the time spent in `acquire()` is what I had in mind.

wddgit commented:
I just pushed more commits and rebased. The Timing service change is included now. @Dr15Jones Do we have a Timing service test running already somewhere?

I think I addressed all the comments at this point. Let me know if there are more.

Each change is in a separate commit. I'd like to squash all this when the review is done.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41001/34765

- This PR adds an extra 132KB to repository

cmsbuild commented:
Pull request #41001 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @makortel can you please check and sign again.

wddgit commented:
please test

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-46a227/31498/summary.html
**COMMIT**: c6a523df7fe97f235db3fae2c372d4f0f0dcfe30
**CMSSW**: CMSSW_13_1_X_2023-03-21-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/41001/31498/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found errors in the following unit tests:

<pre>---> test PrimaryVertex had ERRORS
---> test testPVPlotting had ERRORS
---> test trackerMaterialAnalysisPlots had ERRORS
---> test createDBObjecs had ERRORS
and more ...
</pre>

## Comparison Summary



**Summary**:
 - You potentially added 2 lines to the logs
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3550915
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3550890
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 213 log files, 164 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-03-21-1100+46a227/56173/triggerResults): no differences found

makortel commented:
Thanks @wddgit, this PR looks good now (the failed unit tests fail also in the IB). Could you squash the commits now, and we run final round of tests?

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41001/34774

- This PR adds an extra 8KB to repository

wddgit commented:
please test

Code is unchanged from the previous round of tests except the commits were squashed.

wddgit commented:
I also manually ran an additional test adding Timing Service to the new test added by this PR and adding some debug printout. It looks good. I see an expected 5 acquires and 5 produces with an expected 1 second delay each and 11 other produce calls with an expected tiny time increment. And the total comes out a little over 10 seconds with expected increments to the time at the expected points in time.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-46a227/31515/summary.html
**COMMIT**: 6ac2413e73c3b81e4ec9b8d7b4418bb84d9679ed
**CMSSW**: CMSSW_13_1_X_2023-03-22-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/41001/31515/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially added 15 lines to the logs
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3552750
 - DQMHistoTests: Total failures: 0
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3552728
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 213 log files, 164 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-03-22-1100+46a227/56192/triggerResults): no differences found

makortel commented:
+core

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

perrotta commented:
+1
- Mostly relying on the core signature...
